src/locator/index.js

```diff
@@ -1 +1,249 @@
+    this.#cache = new DigestMap()
+    this.#claimFetched = {
+      index_or_location: new DigestMap(),
+      location: new DigestMap(),
+      standard: new DigestMap()
+    }
+    this.#knownSlices = new DigestMap()
+    this.#knownShards = new DigestMap()
+  }
+
+  /** @param {API.MultihashDigest} digest */
+  async locate (digest) {
+    // get the cached data for this CID (CAR CID & offset)
+    let location = this.#cache.get(digest)
+    if (!location) {
+      // no full cached data -- but perhaps we have the shard already?
+      const knownSlice = this.#knownSlices.get(digest)
+      if (knownSlice) {
+        // read the shard
+        await this.#readShard(digest, knownSlice.shardDigest, knownSlice.position)
+      } else {
+        // nope we don't know anything really here, better read for the digest
+        await this.#readClaims(digest, 'standard')
+        // if we now have and index, read the shard
+        const knownSlice = this.#knownSlices.get(digest)
+        if (knownSlice) {
+          // read the shard
+          await this.#readShard(digest, knownSlice.shardDigest, knownSlice.position)
+        }
+      }
+      // seeing as we just read the index for this CID we _should_ have some
+      // index information for it now.
+      location = this.#cache.get(digest)
+      // if not then, well, it's not found!
+      if (!location) return { error: new NotFoundError(digest) }
+    }
+    return { ok: location }
+  }
+
+  /**
+   *
+   * @param {API.MultihashDigest} digest
+   * @param {ShardDigest} shard
+   * @param {Position} pos
+   * @returns
+   */
+  async #readShard (digest, shard, pos) {
+    let location = this.#getShard(shard)
+    if (!location) {
+      await this.#readClaims(shard, 'location')
+      location = this.#getShard(shard)
+      // if not then, well, it's not found!
+      if (!location) return
+    }
+    this.#cache.set(digest, {
+      digest,
+      site: location.site.map(s => ({
+        location: s.location,
+        range: {
+          offset: (s.range?.offset || 0) + pos[0],
+          length: pos[1]
+        },
+        space: s.space
+      }))
+    })
+  }
+
+  /**
+   *
+   * @param {API.MultihashDigest} shardKey
+   * @returns
+   */
+  #getShard (shardKey) {
+    const knownShard = this.#knownShards.get(shardKey)
+    if (knownShard) {
+      return knownShard
+    }
+    return this.#cache.get(shardKey)
+  }
+
+  /**
+   *
+   * @param {API.MultihashDigest} digest
+   * @param {Kind} kind
+   */
+  async #executeReadClaims (digest, kind) {
+    const result = await this.#client.queryClaims({
+      hashes: [digest],
+      match: this.#spaces && { subject: this.#spaces },
+      kind
+    })
+
+    if (result.error) return
+
+    // process any location claims
+    for (const claim of result.ok.claims.values()) {
+      if (claim.type === 'assert/location') {
+        if (claim.range?.length != null) {
+          addOrSetLocation(this.#cache, contentMultihash(claim), {
```

Comment:
@alanshaw 

FWIW this is a complicated code path cause it applies to times:
1. Indexing service returns a stored location claim that has an range
2. Indexing service uses legacy block index to synthesize a location claim

That's why it's stored in cache -- it might be an actual block location you ultimately return to the top level locate path